lisa/util/shell.py

```diff
@@ -170,6 +173,7 @@ def spawn(
         use_pty: bool = False,
         allow_error: bool = True,
     ) -> Any:
+        self.initialize()
```

Comment:
Whis are these calls necessary in every one of these methods? Can the initialization be moved to the `__init__` method?

Reply:
The ssh session may be broken on anytime. The initialize is used to recover it. So it's try best to make sure it's checked before each run. Maybe I should use decorator to do it, any better idea?

Reply:
What happens when the SSH connection is broken? Does `self._inner_shell` gets set to `None`? Is that why those `assert` are there?

The way I see this class here is that it serves mostly as a proxy to the `_inner_shell` object, converting the type of all `PurePath` arguments to `String` along the way. If the connectivity check were done in the `_inner_shell` object, and its commands could take a `PurePath` object (or if it implemented a `__str__` method that would to the work of `_purepath_to_str()`), then the need for this layer of abstraction is greatly reduced.

Am I understanding this correctly?

Reply:
for assert, it makes mypy, static type checking, happy. Or it will prompt the _inner_shell may be None, don't call its methods.

Your understand is right. The SshShell and LocalShell are duck typing with same method signature. The PurePath object is a system object, and it may be called at any place of user's test code. So it's hard to refine from caller side. The underlying inner shell is a libary.

Reply:
I see.  I usually don't like these kinds of proxy classes, because it's very easy to make a mistake that takes hours to fix.  (For instance, I've spent hours debugging issues because of a silly mistake involving copy/paste and forgetting to rename a variable.)

If there was a way to query what the `_inner_shell` offers as commands (e.g. by prefixing all command methods with a `cmd_` prefix, which works well even for cases where a command name is the same as a reserved Python keyword) and create these methods here automatically during `__new__` or `__init__` time, calls to `initialize()` (which could be renamed to `ensure_connected()`), the `PurePath` conversion, and the method call forwarding could all be automated.

I don't know, however, how it would play with auto-completing in IDEs. (I don't use IDEs.)

Reply:
Thank you, Leandro. it's more clear to me now. I added one refactor in backlog, so that use __new__ or decorator to create a wrapper. It should be able to erase 90% simple wrapper methods.